src/error_diffusion.jl

```diff
@@ -27,8 +27,48 @@ struct ErrorDiffusion{T<:AbstractMatrix} <: AbstractCustomColorDither
 end
 ErrorDiffusion(filter; clamp_error=true) = ErrorDiffusion(filter, clamp_error)
 
-# Error diffusion for general color schemes `cs`.
-function (alg::ErrorDiffusion)(
+function binarydither!(alg::ErrorDiffusion, out::GenericGrayImage, img::GenericGrayImage)
+    # this function does not yet support OffsetArray
+    require_one_based_indexing(img)
+
+    # Change from normalized intensities to Float as error will get added!
+    FT = floattype(eltype(out))
+
+    # eagerly promote to the same type to make loop run faster
+    img = FT.(img)
+    filter = eltype(FT).(alg.filter)
+
+    h, w = size(img)
+    fill!(out, zero(eltype(out)))
+
+    drs = axes(alg.filter, 1)
+    dcs = axes(alg.filter, 2)
+    FT0, FT1 = FT(0), FT(1)
+
+    @inbounds for r in 1:h
+        for c in 1:w
```

Comment:
Sorry I didn't catch this before. This `@inbounds` can be dangerous.

A lot of JuliaImages functions would return `OffsetArray` (e.g., `imrotate`) so you can't assume the axes start from `1` here.

A better practice is to

```julia
@inbounds for r in axes(img, 1)
    for c in axes(img, 2)
        ...
        for dr in drs
            for dc in dcs
                if (r + dr) in axes(img, 1) and (c + dc) in axes(img 2)
                    ...
                end
            end
        end
    end
end
```

Just in case if you're not aware of it, `CartesianIndices` is a very good concept to work with multi-dimensional arrays. The current version actually only works on Matrix. If you haven't tried it yet, I'd like to recommend https://julialang.org/blog/2016/02/iteration/. Otherwise, we might need to restrict the methods to 2-dimensional array only.

Reply:
The problem with iterating over CartesianIndices is that error diffusion requires iterations over rows, not columns. Maybe we could "transpose" the image first? Would this be possible without taking a hit in performance?

Reply:
Ah, I remembered our initial discussion on this. https://github.com/JuliaImages/DitherPunk.jl/pull/10#discussion_r627291733

> Would this be possible without taking a hit in performance?

That would become quite tricky I think... What if we changed the implementation to be along with columns, will that affect the results for real images?

Reply:
It will, but I'm starting to think it would be worth it. 
Maybe we can just add a warning to the docs telling users to `transpose -> dither -> transpose` if they really care about row-first iteration.

Reply:
I guess the difference will mostly be visible on patches of constant color. Row first means the quantization error will be spread "horizontally first" instead of vertically. 
You can see this in the horizontal streaks in the sky here:
![image](https://user-images.githubusercontent.com/20258504/139673051-f80c203c-a76c-48b8-8516-4b20fa41d30a.png)

and quite characteristically in the gradient images:
<img width="810" alt="Screenshot 2021-11-01 at 13 34 26" src="https://user-images.githubusercontent.com/20258504/139673324-b6b85cc6-9e49-4de8-b912-f1739671b635.png">

Reply:
Another option is to do something like:

```julia
R = PermutedDimsArray(CartesianIndices(img), ntuple(i->ndims(img)-i+1, ndims(img)))
for p in R
end
```

but if we loop over `CartesianIndices(img)` I believe it can be faster.

Reply:
I think we can handle the `OffsetArray` stuff and the n-dimensional stuff in the next few PRs. This PR looks awesome already.